Added script, to help gather requests made to aws when running terraform

detnon opened the pull request:
Added helper script and some information in the Readme regarding usage

Commit:
Added script, to help gather requests made to aws when running terraform

0atman commented on cybersecuritytools/aws_permissions_helper/aws_requests.py:
```diff
@@ -0,0 +1,9 @@
+
+with open(sys.argv[1],"r") as data:
+    with open('aws_requests.txt','w') as aws:
+        for s in sorted(set([re.search(pattern,d).group(1) for d in data if re.search(pattern, d)])):
```
m8

0atman replied:
This is very clever, a bit too clever for me to understand. Would you rename to be clearer or give names to the following:
```
pattern
sys.argv[1]
data
aws
s
d
``` 

The problem here is nothing is named, or the names add very little value.

0atman reviewed (changes requested):
first thing's first :ok_hand:

akinnane commented:
Can you either make the whole script a one-liner or break it out into one-simple-thing-per-line?

Commit:
Hopefully improved readability

danjoneslf commented on cybersecuritytools/aws_permissions_helper/README.md:
````diff
@@ -0,0 +1,40 @@
+Firstly, you will need to enable the correct logging level by terraform, before the next step run:
+
+```bash
+export TF_LOG=TRACE
````
Presumably if your pattern is looking for DEBUG log lines you can set `export TF_LOG=DEBUG` and make the log file smaller?

danjoneslf commented on cybersecuritytools/aws_permissions_helper/aws_requests.py:
```diff
@@ -0,0 +1,9 @@
+pattern = "((?<=DEBUG: Request\s)(\w*\W\w*))"
+
+with open(sys.argv[1],"r") as data:
+    with open('aws_requests.txt','w') as aws:
```
Deniz showed me a think where you create the 2 contexts and use a single `with` like 

```
terraform_log = open(sys.argv[1],"r") 
aws_request_log = open('aws_requests.txt','w')

with terraform_log, aws_requests_log: 
    ....
```

something like that anyway.

Commit:
Refactored for readability with the help of @0atman and @akinnane

Commit:
Removed unused variable

Commit:
Added grouped_permissions() to jsonise the services and actions

Commit:
Updated Readme

Commit:
Split doctests onto separate lines to stop flake8 complaining
moved the regex logic into it's own function so it can be typed correctly
updated Readme

Commit:
Fixing Typing issues

Commit:
Changed regex pattern to rawstrings

Commit:
Amended the typing for first_match() & added a docstring

Commit:
Fixed the various typing issues that mypy was showing

akinnane commented on cybersecuritytools/aws_permissions_helper/aws_requests.py:
```diff
@@ -0,0 +1,75 @@
+        aws_requests.write(grouped_permissions(requests))
+
+
+def first_match(pattern: Pattern[str], text: str) -> str:
```
Is this right? Should it really return "No Match" when there isn't a match?

detnon replied:
I wasn't sure what to return here, any suggestions?

Its highly unlikely that anything at this point will not match but probably should have a good way to handle that

akinnane commented on cybersecuritytools/aws_permissions_helper/aws_requests.py:
```diff
@@ -0,0 +1,75 @@
+    for request in requests.splitlines():
+        match_service = first_match(SERVICE, request)
+        match_permission = first_match(PERMISSION, request)
+        if match_service not in grouped_permissions:
+            grouped_permissions[match_service] = []
+            grouped_permissions[match_service].append(match_permission)
+        else:
+            grouped_permissions[match_service].append(match_permission)
```
```suggestion
        grouped_permissions.setdefault(match_service, []).append(match_permission)
```

You can use the `setdefault` method on dict's to either get and return the object if the key is set, or save a default object into the key and return that if the key is not set. That's a more elegant way of expressing the above if block. 

https://docs.python.org/3/library/stdtypes.html#dict

detnon replied:
Nice! Updated to use this

Commit:
Changed if/else to use dict.setdefault()

0atman reviewed (approved):
assuming all the existing comments are ok, sure!
